cpp-client/deephaven/dhclient/src/arrowutil/arrow_array_converter.cc

```diff
@@ -246,6 +496,183 @@ struct ColumnSourceToArrayVisitor final : ColumnSourceVisitor {
     CopyValues<LocalTimeChunk>(source, &builder, [](const LocalTime &o) { return o.Nanos(); });
   }
 
+  struct InnerBuilderMaker {
+    explicit InnerBuilderMaker(const ElementType &element_type) {
+      if (element_type.ListDepth() != 1) {
+        auto message = fmt::format("Expected list_depth 1, got {}", element_type.ListDepth());
+        throw std::runtime_error(DEEPHAVEN_LOCATION_STR(message));
+      }
+
+      switch (element_type.Id()) {
+        case ElementTypeId::kChar: {
+          array_builder_ = std::make_shared<arrow::UInt16Builder>();
+          return;
+        }
+
+        case ElementTypeId::kInt8: {
+          array_builder_ = std::make_shared<arrow::Int8Builder>();
+          return;
+        }
+
+        case ElementTypeId::kInt16: {
+          array_builder_ = std::make_shared<arrow::Int16Builder>();
+          return;
+        }
+
+        case ElementTypeId::kInt32: {
+          array_builder_ = std::make_shared<arrow::Int32Builder>();
+          return;
+        }
+
+        case ElementTypeId::kInt64: {
+          array_builder_ = std::make_shared<arrow::Int64Builder>();
+          return;
+        }
+
+        case ElementTypeId::kFloat: {
+          array_builder_ = std::make_shared<arrow::FloatBuilder>();
+          return;
+        }
+
+        case ElementTypeId::kDouble: {
+          array_builder_ = std::make_shared<arrow::DoubleBuilder>();
+          return;
+        }
+
+        case ElementTypeId::kBool: {
+          array_builder_ = std::make_shared<arrow::BooleanBuilder>();
+          return;
+        }
+
+        case ElementTypeId::kString: {
+          array_builder_ = std::make_shared<arrow::StringBuilder>();
+          return;
+        }
+
+        case ElementTypeId::kTimestamp: {
+          // TODO(kosak): will we pass through non-nano units?
```

Comment:
Sounds like a TODO that needs resolving before merge?

Reply:
Yeah. It was a question to myself and the answer is "no, it's impossible to pass through non-nano units" :-)